packages/plugin-ext/src/main/browser/editor/untitled-resource.ts

```diff
@@ -14,39 +14,103 @@
 export class UntitledResourceResolver implements ResourceResolver {
-    resolve(uri: URI): Resource | Promise<Resource> {
-        if (uri.scheme === Schemes.UNTITLED) {
-            return resources.get(uri.toString())!;
+
+    @inject(FileResourceResolver)
+    protected readonly fileResourceResolver: FileResourceResolver;
+
+    @inject(OpenerService)
+    protected readonly openerService: OpenerService;
+
+    async resolve(uri: URI): Promise<UntitledResource> {
+        if (uri.scheme !== Schemes.UNTITLED) {
+            throw new Error('The given uri is not untitled file uri: ' + uri);
+        } else {
+            const untitledResource = resources.get(uri.toString());
+            if (!untitledResource) {
+                return createUntitledResource('', '', uri, this.fileResourceResolver);
+            } else {
+                return untitledResource;
+            }
         }
-        throw new Error(`scheme ${uri.scheme} is not '${Schemes.UNTITLED}'`);
     }
 }
 
 export class UntitledResource implements Resource {
+    private fileResource: FileResource;
 
-    constructor(public uri: URI, private content?: string) {
+    constructor(public uri: URI, private content?: string, fileResourceResolver?: FileResourceResolver) {
+        const that = this;
         resources.set(this.uri.toString(), this);
+        if (fileResourceResolver) {
+            fileResourceResolver.resolveWithoutSchemaCheck(uri).then(resolvedFileResource => {
+                that.fileResource = resolvedFileResource;
+            });
+        }
     }
 
-    readContents(options?: { encoding?: string | undefined; } | undefined): Promise<string> {
-        return Promise.resolve(this.content ? this.content : '');
+    async init(): Promise<void> {
+        if (this.fileResource) {
+            return this.fileResource.init();
+        }
     }
 
     dispose(): void {
         resources.delete(this.uri.toString());
     }
+
+    async readContents(options?: { encoding?: string | undefined; } | undefined): Promise<string> {
+        if (this.content) {
+            return Promise.resolve(this.content);
+        } else if (this.fileResource) {
+            return this.fileResource.readContents(options);
+        } else {
+            return Promise.resolve('');
+        }
+    }
+
+    async saveContents(content: string, options?: { encoding?: string, overwriteEncoding?: string }): Promise<void> {
+        if (this.fileResource) {
+            this.fileResource.saveContents(content, options);
+        }
+    }
+
+    async saveContentChanges(changes: TextDocumentContentChangeEvent[], options?: { encoding?: string, overwriteEncoding?: string }): Promise<void> {
+        if (this.fileResource) {
+            this.fileResource.saveContentChanges(changes, options);
+        }
+    }
+}
+
+export async function createUntitledResource(content?: string, language?: string, uri?: URI, fileResourceResolver?: FileResourceResolver): Promise<UntitledResource> {
```

Comment:
global variables and functions is anti-pattern in Theia because of extensibility, we should get rid fo them, please see https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#di-function-export